org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java

```diff
@@ -6099,7 +6103,13 @@ protected int resumeAfterRecovery() {
 	if (this.assistNode != null) {
 
 		if (requireExtendedRecovery()) {
-			if (this.unstackedAct != ERROR_ACTION) {
+			if (this.unstackedAct != ERROR_ACTION
+					// resume also if we are inside a MessageSend to complete collecting all paramaters
+					// when using generics, it helps for proper resolution of the MessageSend when all
+					// paramaters are collected. This is essential when next few paramaters are the ones
+					// that decide a TypeParameter of a current completing paramater type.
+					|| (topKnownElementKind(COMPLETION_OR_ASSIST_PARSER) == K_BETWEEN_CASE_AND_COLON
```

Comment:
To be honest I'm not proud of this condition, it's just trail and error to get the parser to parts the whole invocation so that the lamda variable is resolved to have proper completions inside switch expressions. Thats one reason i want to get feedback from you, because I'm sure there should be a more elegant way to handle this.

I do agree, if we generlize this behavior or could generalize it, so that when we are completing on a method invocation which has type parameters, if we could continue until the invocation expression ends, it could benifit other cases as well isn't it ?

Reply:
> Seeing this code change under the condition of `requireExtendedRecovery()` I wonder if all this is indeed specific to the situation of a switch expression inside a lambda (which it what requireExtendedRecovery tests)? Or would the same tweaks also be necessary when the switch itself (i.e., its value) is passed into a method invocation?

No for normal method invocation, the current implementation works. Isn't only when inside lambda.